### PR TITLE
Fix #542, adblock stats counting improvements.

### LINKED ISSUE #542: Use adblock lib for more precise shield block counting.

### PR DESCRIPTION
**Pull request by iccub**
Please check it out and see if it compiles and show blocked stats correctly for you.

I took `Adblocker.swift` class from 1.6 and cleaned it up a bit. 

Notes:
- I added support of downloading regional filters, although I don't see it in the app, do we even use region specific adblocking?
- I updated adblock library dependency to the latest version and pinned it.
- Some websites under or overreport number of blocked ads compared to 1.6

## Pull Request Checklist

- [ ] My patch has gone through review and I have addressed review comments
- [x] My patch has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!`

- [ ] I have updated the *Unit Tests* to cover new or changed functionality
- [ ] I have updated the *UI Tests* to cover new or changed functionality
- [ ] I have marked the bug with `[needsuplift]`
- [ ] I have made sure that localizable strings use `NSLocalizableString()`

## Screenshots

If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here.

## Notes for testing this patch

If useful, please leave notes for QA, explaining what this patch changes and how it can be best tested and verified.

### CODE REVIEW

File: Client/Frontend/ContentBlocker/TrackingProtectionPageStats.swift [hunk 1]
```diff
@@ -75,11 +75,10 @@ class TPStatsBlocklistChecker {
                     }
                 }
             }
-            if let blocklist = blockLists.blocklistForURL(url), enabledLists.contains(blocklist) {
-                deferred.fill(blocklist)
-            } else {
-                deferred.fill(nil)
-            }
+            
+            let shouldBlockRequest = enabledLists.contains(.ad) && AdBlockStats.singleton.shouldBlock(request)
```
Comment (kylehickinson) [inline Client/Frontend/ContentBlocker/TrackingProtectionPageStats.swift]: What about `BlocklistName.trackers`?

File: Client/WebFilters/AdblockStats/AdBlockStats.swift [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+import Foundation
+import Shared
+import BraveShared
+
+private let log = Logger.browserLogger
+
+class AdBlockStats {
+    static let singleton = AdBlockStats()
+    
+    typealias localeCode = String
```
Comment (kylehickinson) [inline Client/WebFilters/AdblockStats/AdBlockStats.swift]: nit: `localCode` → `LocalCode`

File: Client/WebFilters/AdblockStats/AdBlockStats.swift [hunk 3]
```diff
@@ -0,0 +1,242 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Shared
+import BraveShared
+
+private let log = Logger.browserLogger
+
+class AdBlockStats {
+    static let singleton = AdBlockStats()
```
Comment (kylehickinson) [inline Client/WebFilters/AdblockStats/AdBlockStats.swift]: nit: `singleton` → `shared`

File: Client/WebFilters/AdblockStats/AdBlockStats.swift [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+            return url
+        }
+    }
+}
+
+extension AdBlockStats: NetworkDataFileLoaderDelegate {
+    
+    func fileLoader(_ loader: NetworkDataFileLoader, setDataFile data: Data?) {
+        guard let loader = loader as? AdblockNetworkDataFileLoader, let adblocker = abpFilterLibWrappers[loader.lang] else {
+            assert(false)
```
Comment (kylehickinson) [inline Client/WebFilters/AdblockStats/AdBlockStats.swift]: nit: `assert(false)` → `assertionFailure()`

File: Client/WebFilters/AdblockStats/AdBlockStats.swift [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+            return
+        }
+        adblocker.setDataFile(data)
+    }
+    
+    func fileLoaderHasDataFile(_ loader: NetworkDataFileLoader) -> Bool {
+        guard let loader = loader as? AdblockNetworkDataFileLoader, let adblocker = abpFilterLibWrappers[loader.lang] else {
+            assert(false)
```
Comment (kylehickinson) [inline Client/WebFilters/AdblockStats/AdBlockStats.swift]: Same as above

File: Client/WebFilters/AdblockStats/NetworkDataFileLoader.swift [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        if !delegate.fileLoaderHasDataFile(self) {
+            networkRequest()
+        } else if !delegate.fileLoaderDelegateWillHandleInitialRead(self) {
+            func readData() -> Data? {
+                guard let path = pathToExistingDataOnDisk() else { return nil }
+                return FileManager.default.contents(atPath: path)
+            }
+            
+            let data = readData()
+            if data == nil {
```
Comment (kylehickinson) [inline Client/WebFilters/AdblockStats/NetworkDataFileLoader.swift]: Why not just 

```swift
if let data = readData() {
  delegate.fileLoader(self, setDataFile: data)
} else {
  networkRequest()
}
```

**Review by kylehickinson — CHANGES_REQUESTED** [review]

**Comment by kylehickinson** [comment]
Just some minor nit's :) Realize most of this is probably ported

File: package.json [hunk 7]
```diff
@@ -14,7 +15,8 @@
   "dependencies": {
     "page-metadata-parser": "^1.1.2",
     "readability": "mozilla/readability#3c76104adb00eb9897517e6be6c74c6c05918dfe",
-    "brave-sync": "brave/sync#staging"
+    "brave-sync": "brave/sync#staging",
+    "ad-block": "git+ssh://git@github.com/brave/ad-block.git#dcfc5847b6291d30ac07537cb861772c1a158ece"
```
Comment (kylehickinson) [inline package.json]: I think this line is tripping up Travis. May need to be a non-ssh URL

File: Client/Frontend/ContentBlocker/TrackingProtectionPageStats.swift (hunk 1 above)
Comment (iccub) [reply Client/Frontend/ContentBlocker/TrackingProtectionPageStats.swift]: I don't think Adblock lib supports trackers, cc @jhreis to confirm

Comment (jhreis) [reply Client/Frontend/ContentBlocker/TrackingProtectionPageStats.swift]: Yeah, it supports tracking protection, however, our lists are combined, so probably not going to really matter.

**Comment by kylehickinson** [comment]
"Approved" -- Just waiting on travis fixes

**Comment by iccub** [comment]
Finally managed to fix failing travis builds. Please rebase and merge once approved.

**Review by kylehickinson — APPROVED** [review]

**Comment by kylehickinson** [comment]
Looks like it needs another development merge/rebase

**Comment by iccub** [comment]
Yep, I will wait for @jhreis  to say few words about this PR then clean it up before merging.

File: Shared/Extensions/StringExtensions.swift [hunk 8]
```diff
@@ -96,4 +96,16 @@ public extension String {
         let cleaned = self.replacingOccurrences(of: "\n", with: " ")
         return cleaned.trimmingCharacters(in: .whitespacesAndNewlines).components(separatedBy: string)
     }
+    
+    public func startsWith(_ other: String) -> Bool {
```
Comment (jhreis) [inline Shared/Extensions/StringExtensions.swift]: Can't we just use: https://developer.apple.com/documentation/swift/string/2894028-starts

**Review by jhreis — CHANGES_REQUESTED** [review]

File: Shared/Extensions/StringExtensions.swift (hunk 8 above)
Comment (iccub) [reply Shared/Extensions/StringExtensions.swift]: We don't even use it, removed.

**Review by jhreis — APPROVED** [review]

**Pull request merged**